src/edu/cmu/cs/sasylf/ast/QualName.java

```diff
@@ -19,11 +23,53 @@
  * </ul>
  */
 public class QualName extends Node {
-	private final QualName source;
-	private final String name;
+	private QualName source;
+	private String name;
+
+	/*
+	 * Substitutable indicates whether this QualName is allowed to be replaced during substitution.
+	 * Qualnames can be replaced during substitution in ModulePart.substitute.
+	 * Substitutable is set to true by default, and should be set to false after the QualName is substituted into.
+	 * This is to prevent the QualName from being replaced if there is a QualName module parameter with the same symbol.
+	 */
+	private boolean substitutable; // whether this QualName is allowed to be replaced during substitution
 	
 	private Object resolution;
 	private int version;
+
+	public String getName() {
+		return name;
+	}
+	
+	/**
+	 * Whether this QualName is allowed to be replaced during substitution.
+	 * @return
+	 */
+	public boolean isSubstitutable() {
+		return substitutable;
+	}
+
+	/**
+	 * Set this QualName to be unsubstitutable.
+	 * Should be called after a QualName is substituted into.
+	 */
+	public void setUnsubstitutable() {
+		substitutable = false;
+	}
+
+	/**
+	 * Set the resolution of this QualName to null.
+	 * 
+	 * This is used after cloning and substitution so that the resolution is
+	 * recomputed when needed.
+	 */
+	public void nullifyResolution() {
+		resolution = null;
+	}
+
+	public void setName(String newName) {
```

Comment:
`setName()` was being used for substitution in `ModulePart`. I have refactored the code so that `QualName` substitution occurs in `QualName.substitute()`, which allowed me to get rid of `setName()` and make `name` private.